src/docfx/build/xref/XrefResolver.cs

```diff
@@ -58,76 +58,58 @@ internal class XrefResolver
-            {
-                queries["view"] = moniker;
-            }
-
-            var resolvedHref = UrlUtility.MergeUrl(
-                RemoveSharingHost(xrefSpec.Href, _config.HostName),
-                queries.AllKeys.Length == 0 ? "" : "?" + string.Join('&', queries),
-                fragment.Length == 0 ? "" : fragment);
-
-            // NOTE: this should also be relative to root file
-            var sourceFile = inclusionRoot ?? hrefRelativeTo;
-            _fileLinkMapBuilder.AddFileLink(sourceFile.FilePath, sourceFile.SiteUrl, resolvedHref);
-
-            if (xrefSpec?.DeclaringFile != null && inclusionRoot != null)
-            {
-                resolvedHref = UrlUtility.GetRelativeUrl(inclusionRoot.SiteUrl, resolvedHref);
-            }
-
             return (null, resolvedHref, display, xrefSpec?.DeclaringFile);
         }
 
-        public (Error?, ExternalXrefSpec?) ResolveXrefSpec(SourceInfo<string> uid, Document referencingFile)
+        public (Error?, ExternalXrefSpec?, string? href) ResolveXrefSpec(SourceInfo<string> uid, Document referencingFile)
         {
-            var (error, xrefSpec) = Resolve(uid, referencingFile);
-            return (error, xrefSpec?.ToExternalXrefSpec());
+            var (error, xrefSpec, href) = Resolve(uid, referencingFile, referencingFile);
+            if (xrefSpec == null)
+                return (error, null, null);
+
+            return (error, xrefSpec.ToExternalXrefSpec(), href);
         }
 
         public XrefMapModel ToXrefMapModel()
         {
-            string? repositoryBranch = null;
-            string? basePath = null;
+            var repositoryBranch = _repository?.Branch;
+            var basePath = _config.BasePath.ValueWithLeadingSlash;
+
             var references = _internalXrefMap.Value.Values
                 .Select(xref =>
                 {
-                    var xrefSpec = xref.ToExternalXrefSpec();
-                    if (repositoryBranch is null)
-                    {
-                        repositoryBranch = _repository?.Branch;
-                    }
-                    if (basePath is null)
-                    {
-                        basePath = _config.BasePath.ValueWithLeadingSlash;
-                    }
-
                     // DHS appends branch infomation from cookie cache to URL, which is wrong for UID resolved URL
                     // output xref map with URL appending "?branch=master" for master branch
-                    var (_, _, fragment) = UrlUtility.SplitUrl(xref.Href);
-                    var path = $"https://{_xrefHostName}{xref.DeclaringFile.SiteUrl}";
                     var query = repositoryBranch == "master" ? "?branch=master" : "";
-                    xrefSpec.Href = UrlUtility.MergeUrl(path, query, fragment);
+                    var href = UrlUtility.MergeUrl($"https://{_xrefHostName}{xref.Href}", query);
+
+                    var xrefSpec = xref.ToExternalXrefSpec(href);
```

Comment:
If we keep the public setter in `Href` property, we can get rid of this `overwriteHref`?

Reply:
Prefer private setter to avoid unintentional assign. 
I just wasted a long time local debuging a weird behavior. Orzzzzz

Reply:
IMHO, this looks weird as well... what kind of unintentional assignment did you get?

Reply:
A multi-threading bug caused by shared state mutation. Our best practice is to use _immutable_, _readonly_ object model as much as you can.

Reply:
Agree with the best practice, my point is we are mutating it anyway.